ui/Modal/Org/ConfigureEns/RegisterName.svelte

```diff
@@ -0,0 +1,221 @@
+  import * as mutexExecutor from "ui/src/mutexExecutor";
+  import * as svelteStore from "ui/src/svelteStore";
+  import * as validation from "ui/src/validation";
+  import * as wallet from "ui/src/wallet";
+
+  import { Modal, TextInput } from "ui/DesignSystem";
+
+  import ButtonRow from "./ButtonRow.svelte";
+  import ConfirmRegistration from "./ConfirmRegistration.svelte";
+
+  export let registrationDone: (result: Result) => void;
+  export let currentName: string | undefined;
+  export let fee: ethers.BigNumber;
+
+  let nameInputValue: string = currentName || "";
+
+  type State =
+    | {
+        type: "validateAndCommit";
+      }
+    | {
+        type: "register";
+        commitmentSalt: Uint8Array;
+        commitmentBlock: number;
+        minAge: number;
+      };
+
+  let state: State = { type: "validateAndCommit" };
+
+  let userInputStarted: boolean = nameInputValue !== "";
+  let validationState: validation.ValidationState = {
+    status: validation.ValidationStatus.NotStarted,
+  };
+  let registration: ensResolver.Registration | null;
+
+  $: validateName(nameInputValue);
+
+  function validateName(nameInputValue: string | undefined): void {
+    if (!userInputStarted) {
+      userInputStarted = true;
+      return;
+    }
+
+    if (!nameInputValue) {
+      validationState = {
+        status: validation.ValidationStatus.Error,
+        message: "You need to enter a name.",
+      };
+    } else {
+      validationState = {
+        status: validation.ValidationStatus.Loading,
+      };
+      debouncedValidateFormAndSetState();
+    }
+  }
+
+  const debouncedValidateFormAndSetState = lodash.debounce(
+    validateFormAndSetState,
+    1000
+  );
+
+  const validateFormExecutor = mutexExecutor.create();
+
+  async function validateFormAndSetState(): Promise<void> {
+    const validationResult = await validateFormExecutor.run(async () => {
+      return await validateForm();
+    });
+
+    if (validationResult) {
+      ({ validationState, registration } = validationResult);
+    }
+  }
+
+  async function validateForm(): Promise<{
+    validationState: validation.ValidationState;
+    registration: ensResolver.Registration | null;
+  }> {
+    const available = await ensRegistrar.isAvailable(nameInputValue);
+
+    if (available) {
+      const accountBalancesStore = wallet.accountBalancesStore;
+      const radBalance = svelteStore.get(accountBalancesStore).rad;
+
+      if (radBalance && radBalance < fee) {
+        return {
+          validationState: {
+            status: validation.ValidationStatus.Error,
+            message:
+              "You don't have enough RAD in your wallet to register this name.",
+          },
+          registration: null,
+        };
+      }
+
+      return {
+        validationState: {
+          status: validation.ValidationStatus.Success,
+        },
+        registration: null,
+      };
+    }
+
+    registration = await ensResolver.getRegistration(
```

Comment:
This mutates the component state variable and leads to a race.
```suggestion
    const registration = await ensResolver.getRegistration(
```

Reply:
Good catch!